xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/XWikiAction.java

```diff
@@ -731,45 +732,30 @@ public void execute(XWikiContext context) throws Exception
-     * If the action is get, it should returns a 404 if the sheet parameter is not used (or if the given sheet doesn't
-     * exist either) and if the disableCheckNotExisting parameter is not used.
+     * Check if the given document exists or not and if it should return a 404 based on the context. A {@link
+     * DocExistValidator} with an hint matching the current action is used to check if the document exists. When no
+     * {@link DocExistValidator} is found, the response is always {@code false} When a {@link DocExistValidator} is
+     * found, the result is delegated to {@link DocExistValidator#docExist(XWikiDocument, XWikiContext)}.
      *
-     * @param doc the doc for which to check it exists or not.
+     * @param doc the doc for which to check it exists or not
      * @param context the current context
      * @return {@code true} if we should return a 404
      */
     private boolean shouldReturnDocDoesNotExist(XWikiDocument doc, XWikiContext context)
     {
         boolean result = false;
-        if (doc.isNew()) {
-            String action = context.getAction();
-            XWikiRequest request = context.getRequest();
-            if ("view".equals(action)) {
-                result = !"recyclebin".equals(request.get("viewer"))
-                    && !"children".equals(request.get("viewer"))
-                    && !"siblings".equals(request.get("viewer"));
-            } else if ("get".equals(action)) {
-                String sheet = request.get("sheet");
-                if (!StringUtils.isEmpty(sheet)) {
-                    DocumentReference sheetReference = getCurrentMixedDocumentReferenceResolver().resolve(sheet);
-                    try {
-                        XWikiDocument sheetDoc = context.getWiki().getDocument(sheetReference, context);
-                        result = sheetDoc.isNew();
-                    } catch (XWikiException e) {
-                        LOGGER.warn("Error while trying to load sheet [{}] for checking status code "
-                            + "on GET request for [{}]: [{}]", sheetReference, doc.getDocumentReference(),
-                            ExceptionUtils.getRootCauseMessage(e));
-                        // there is an error we consider that the sheet doesn't exist.
-                        result = true;
-                    }
-                } else {
-                    result = !"1".equals(request.get("disableCheckNotExisting"));
-                }
+        String action = context.getAction();
+        try {
+            DocExistValidator instance = this.componentManager.getInstance(DocExistValidator.class, action);
+            if (instance != null) {
+                result = instance.docExist(doc, context);
             }
+        } catch (ComponentLookupException e) {
+            // It's not required to have a validator for each action, when not specific component is provided for 
+            // an action, the default false value is returned. 
```

Comment:
Not sure about this fallback, I'd bulletproof a bit by checking if we're not in get/view and letting the exception propagate in such case.

Reply:
Why would we hard-code get/view specifically?

Reply:
I was mentioning those since that's the one for which we do provide components implem in XS: we know that those instances should exist, so if we don't find them it's really an error.

Reply:
ok, make sense, thanks.